x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/rank/textsimilarity/TextSimilarityRankFeaturePhaseRankCoordinatorContext.java

```diff
@@ -51,6 +56,22 @@ public TextSimilarityRankFeaturePhaseRankCoordinatorContext(
 
     @Override
     protected void computeScores(RankFeatureDoc[] featureDocs, ActionListener<float[]> scoreListener) {
+        // The rerank inference endpoint may have an override to return top N documents only, in that case let's fail fast to avoid
+        // assigning scores to the wrong inputs
+        Integer inferenceEndpointTopN = getTopNFromInferenceEndpointTaskSettings();
+        if (inferenceEndpointTopN != null && inferenceEndpointTopN < featureDocs.length) {
+            scoreListener.onFailure(
+                new IllegalArgumentException(
+                    "Inference endpoint ["
+                        + inferenceId
+                        + "] is configured to return the top ["
+                        + inferenceEndpointTopN
+                        + "] results. Reduce rank_window_size to be less than or equal to this value."
+                )
+            );
+            return;
+        }
```

Comment:
We should definitely keep this asynchronous. Rule of thumb, never use an ActionFuture or make a async call synchronous when you can easily avoid it. 

It should be something like this (obviously with things better filled in)
```java
        // The rerank inference endpoint may have an override to return top N documents only, in that case let's fail fast to avoid
        // assigning scores to the wrong inputs
        // rank Listener
        final ActionListener<InferenceAction.Response> actionListener = scoreListener.delegateFailureAndWrap((l, r) -> {
            // rank task
        }); 

        // top N listener
        GetInferenceModelAction.Request request = new GetInferenceModelAction.Request(inferenceId, TaskType.RERANK, );
        ActionListener<GetInferenceModelAction.Response> topNListener = ActionListener.wrap(
            topN -> {
                Integer topNDocs = null;
                if (topN.getEndpoints().get(0).getTaskSettings() instanceof CohereRerankTaskSettings cohere) {
                    topNDocs = cohere.getTopNDocumentsOnly();
                } else if (topN.getEndpoints().get(0).getTaskSettings() instanceof GoogleVertexAiRerankTaskSettings googl) {
                    topNDocs = googl.topN();
                }
                if (topNDocs != null && topNDocs < featureDocs.length) {
                    scoreListener.onFailure(
                        new IllegalArgumentException(
                            "Inference endpoint ["
                                + inferenceId
                                + "] is configured to return the top ["
                                + topNDocs 
                                // indicate what the rank window size is
                                + "] results. Reduce rank_window_size to be less than or equal to this value."
                        )
                    );
                    return;
                }
                List<String> featureData = Arrays.stream(featureDocs).map(x -> x.featureData).toList();
                InferenceAction.Request inferenceRequest = generateRequest(featureData);
                try {
                    client.execute(InferenceAction.INSTANCE, inferenceRequest, actionListener);
                } finally {
                    request.decRef();
                }
            },
            f -> {
                // throw appropriate errors.
            }
        );
client.execute(GetInferenceModelAction.INSTANCE, request, topNListener);
```

Reply:
@benwtrent Thanks for the code snippet, I updated it.

One question regarding your comment:
```
            f -> {
                // throw appropriate errors.
            }
```
Isn't it sufficient to just delegate to the outer listener, i.e. `scoreListener.onFailure(f)`?